File: app/src/main/java/br/com/example/hamburgos/request/Repository.java
package br.com.example.hamburgos.request;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import br.com.example.hamburgos.BuildConfig;
import br.com.example.hamburgos.model.Ingredient;
import br.com.example.hamburgos.model.Snack;
import br.com.example.hamburgos.util.Constants;
import br.com.example.hamburgos.util.Utils;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Repository {

    private Context context;
    private Service service;

    public Repository(Context context) {

        this.context = context;

        //setup cache
        File httpCacheDirectory = new File(context.getCacheDir(), "responses");
        int cacheSize = 10 * 1024 * 1024; // 10 MiB
        Cache cache = new Cache(httpCacheDirectory, cacheSize);

        final Interceptor interceptor = new Interceptor() {
            @Override
            public Response intercept(Interceptor.Chain chain) throws IOException {
                Response originalResponse = chain.proceed(chain.request());
                if (Utils.isNetworkAvailable(Repository.this.context)) {
                    int maxAge = 60; // read from cache for 1 minute
                    return originalResponse.newBuilder()
                            .header("Cache-Control", "public, max-age=" + maxAge)
                            .build();
                } else {
                    int maxStale = 60 * 60 * 24 * 28; // tolerate 4-weeks stale
                    return originalResponse.newBuilder()
                            .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                            .build();
                }
            }
        };

        OkHttpClient client = new OkHttpClient.Builder()
                .cache(cache)
                .addInterceptor(interceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        service = retrofit.create(Service.class);
    }

    private Map<String, String> createParans() {
        Map<String, String> map = new HashMap<>();
        map.put(Constants.KEY, BuildConfig.API_KEY);
//        map.put(Constants.QUERY, query);

        return map;
    }

    public void listSnacks(Callback<List<Snack>> callback) {
        Call<List<Snack>> call = service.listSnacks();
        call.enqueue(callback);
    }

    public void listIngridients(Callback<List<Ingredient>> callback) {
        Call<List<Ingredient>> call = service.listIngredients();
        call.enqueue(callback);
    }

}

File: app/src/main/java/br/com/example/hamburgos/request/Repository.kt
package br.com.example.hamburgos.request

import br.com.example.hamburgos.model.Ingredient
import br.com.example.hamburgos.model.Order
import br.com.example.hamburgos.model.Promotion
import br.com.example.hamburgos.model.Snack
import br.com.example.hamburgos.util.Constants
import retrofit2.Callback
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class Repository internal constructor() {

    private val service: Service

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        service = retrofit.create(Service::class.java)
    }

    internal fun listSnacks(callback: Callback<List<Snack>>) {
        val call = service.listSnacks()
        call.enqueue(callback)
    }

    internal fun getSnackById(snackId: Int, callback: Callback<Snack>) {
        val call = service.getSnackById(snackId)
        call.enqueue(callback)
    }

    internal fun getIngredientBySnack(snackId: Int, callback: Callback<List<Ingredient>>) {
        val call = service.getIngredientsBySnack(snackId)
        call.enqueue(callback)
    }

    internal fun getIngredientById(id: Int, callback: Callback<Ingredient>) {
        val call = service.getIngredientsById(id)
        call.enqueue(callback)
    }

    internal fun listIngredients(callback: Callback<List<Ingredient>>) {
        val call = service.listIngredients()
        call.enqueue(callback)
    }

    internal fun listPromotions(callback: Callback<List<Promotion>>) {
        val call = service.listPromotions()
        call.enqueue(callback)
    }

    internal fun listOrders(callback: Callback<List<Order>>) {
        val call = service.listOrders()
        call.enqueue(callback)
    }

    internal fun addOrder(snack: Snack, callback: Callback<Order>) {
        val call = service.addOrder(snack.id, snack.jsonExtras)
        call.enqueue(callback)
    }

}
